activation/nipost.go

```diff
@@ -222,5 +195,148 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.Hash3
 		NIPost: &types.NIPost{},
 	}
 	nb.persist()
-	return nipost, nil
+	return nipost, postGenDuration, nil
+}
+
+// Submit the challenge to a single PoET.
+func submitPoetChallenge(ctx context.Context, logger log.Log, poet PoetProvingServiceClient, challenge *types.Hash32) (*PoetRequest, error) {
+	poetServiceID, err := poet.PoetServiceID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get PoET service ID: %v", ErrPoetServiceUnstable, err)
+	}
+
+	logger.With().Debug("submitting challenge to poet proving service",
+		log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+		log.Stringer("challenge", *challenge))
+
+	round, err := poet.Submit(ctx, *challenge)
+	if err != nil {
+		logger.With().Error("failed to submit challenge to poet proving service",
+			log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+			log.Stringer("challenge", *challenge),
+			log.Err(err))
+		return nil, fmt.Errorf("%w: failed to submit challenge to poet service: %v", ErrPoetServiceUnstable, err)
+	}
+
+	logger.With().Info("challenge submitted to poet proving service",
+		log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+		log.String("round_id", round.ID),
+		log.Stringer("challenge", *challenge))
+
+	return &PoetRequest{
+		PoetRound:     round,
+		PoetServiceID: poetServiceID,
+	}, nil
+}
+
+// Submit the challenge to all registered PoETs.
+func (nb *NIPostBuilder) submitPoetChallenges(ctx context.Context, challenge *types.Hash32) []PoetRequest {
+	g, ctx := errgroup.WithContext(ctx)
+	poetRequestsChannel := make(chan PoetRequest)
+	for _, poetProver := range nb.poetProvers {
+		poet := poetProver
+		g.Go(func() error {
+			if poetRequest, err := submitPoetChallenge(ctx, nb.log, poet, challenge); err == nil {
+				poetRequestsChannel <- *poetRequest
+			} else {
+				nb.log.With().Warning("failed to submit challenge to PoET", log.Err(err))
+			}
+			return nil
+		})
+	}
+	poetRequests := make([]PoetRequest, 0)
+	done := make(chan struct{})
+	go func() {
+		for request := range poetRequestsChannel {
+			poetRequests = append(poetRequests, request)
+		}
+		close(done)
+	}()
+	g.Wait()
+	close(poetRequestsChannel)
+	<-done
+	return poetRequests
+}
```

Comment:
```suggestion
// Submit the challenge to all registered PoETs.
func (nb *NIPostBuilder) submitPoetChallenges(ctx context.Context, challenge *types.Hash32) []PoetRequest {
	poetRequestsChannel := make(chan PoetRequest, len(nb.poetProvers))
	poetRequests := make([]PoetRequest, 0, len(nb.poetProvers))

	wg := sync.WaitGroup{}
	for _, poetProver := range nb.poetProvers {
		poet := poetProver
		wg.Add(1)
		go func() {
			if poetRequest, err := submitPoetChallenge(ctx, nb.log, poet, challenge); err == nil {
				poetRequestsChannel <- *poetRequest
			} else {
				nb.log.With().Warning("failed to submit challenge to PoET", log.Err(err))
			}
			wg.Done()
		}()
	}
	wg.Wait()
	close(poetRequestsChannel)

	for request := range poetRequestsChannel {
		poetRequests = append(poetRequests, request)
	}
	return poetRequests
}
```

I completely forgot one can still read elements from a channel after it is closed (as long as it is a buffered channel).

Reply:
It is simpler but it has the drawback that the consumer starts processing after all producers are finished. In this case, it doesn't really matter as it just collects the results and I don't expect too many PoETs so I will refactor.

However, I'd rather keep the current implementation of `awaitPoetProof`, which uses the same approach as here.